components/lib/auth/src/main/java/mozilla/components/lib/auth/BiometricUtils.kt

```diff
@@ -0,0 +1,54 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.auth
+
+import android.content.Context
+import android.os.Build
+import androidx.annotation.VisibleForTesting
+import androidx.biometric.BiometricManager
+
+/**
+ * Utility class for BiometricPromptAuth
+ */
+class BiometricUtils {
```

Comment:
Replying back to https://github.com/mozilla-mobile/android-components/pull/12291#discussion_r896718665= over here as well:

Ah, I didn't catch this earlier. I see that you making the utils class an instance for testing. Consider now though, that if you want to use `canUseFeature` in your application code, you now have to create an instance of this class every time.

```kotlin
// Expanded into two lines for clarity.
val utils = BiometricUtils()
utils.canUseFeature(context)

// A better API would not need an instance if it was static.
BiometricUtils.canUseFeature(context)
```

So we've made testing "easier" but it's at the cost of bad design pattern. You have tests for `isHardwareAvailable` and `isEnrolled` which is where our logic is and the things we should be testing. The last bit of logic that you would want to test is the bit in `canUseFeature`:

```kotlin
return isHardwareAvailable(manager) && isEnrolled(manager)
```

We can write this code in a way so that we pass the manager to it in an internal function (that can be mocked and tested just like you did for the other methods, then leave the Android platform code as a wrapper. What that would look like:

```kotlin
// Our public API for the extension function. Only test to write is for the SDK_INT case when it's false.
fun Context.canUseBiometricFeature(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
        val manager = BiometricManager.from(this)
        return BiometricUtils.canUseFeature(manager)
    } else {
        false
    }
}

internal object BiometricUtils {

    // Keep our logic in a static function that takes in it's dependencies (here, it's the BiometricManager) as a parameter which we can mock)
    internal fun canUseFeature(manager: BiometricManager): Boolean {
        return isHardwareAvailable(manager) && isEnrolled(manager)
    }

    internal fun isHardwareAvailable(manager: BiometricManager): Boolean {}

    internal fun isEnrolled(manager: BiometricManager): Boolean {}
}
```

Let me know if that makes sense or not. Happy to talk about it more. 🙂

Reply:
Done